internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,313 @@
+		ok      bool
+	)
+	dataMap, err = p.bufferBatchGetter(ctx, [][]byte{k})
+	if err != nil {
+		return nil, err
+	}
+	v, ok = dataMap[string(k)]
+	if !ok {
+		return nil, tikverr.ErrNotExist
+	}
+	return v, nil
+}
+
+func (p *PipelinedMemDB) GetFlags(k []byte) (kv.KeyFlags, error) {
+	f, err := p.memBuffer.GetFlags(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		f, err = p.flushingMemBuffer.GetFlags(k)
+	}
+	if err != nil {
+		return 0, err
+	}
+	return f, nil
+}
+
+func (p *PipelinedMemDB) UpdateFlags(k []byte, ops ...kv.FlagsOp) {
+	p.memBuffer.UpdateFlags(k, ops...)
+}
+
+func (p *PipelinedMemDB) RemoveFromBuffer(key []byte) {}
+
+// Set sets the value for key k in the MemBuffer.
+func (p *PipelinedMemDB) Set(key, value []byte) error {
+	return p.memBuffer.Set(key, value)
+}
+
+// SetWithFlags sets the value for key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) SetWithFlags(key, value []byte, ops ...kv.FlagsOp) error {
+	return p.memBuffer.SetWithFlags(key, value, ops...)
+}
+
+// Delete deletes the key k in the MemBuffer.
+func (p *PipelinedMemDB) Delete(key []byte) error {
+	return p.memBuffer.Delete(key)
+}
+
+// DeleteWithFlags deletes the key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) DeleteWithFlags(key []byte, ops ...kv.FlagsOp) error {
+	return p.memBuffer.DeleteWithFlags(key, ops...)
+}
+
+// Flush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemBuffer is done.
+// When the mutable memdb is too large, it blocks until the ongoing flush is done.
+func (p *PipelinedMemDB) Flush(force bool) (bool, error) {
+	if p.flushFunc == nil {
+		return false, errors.New("flushFunc is not provided")
+	}
+	if !force && !p.needFlush() {
+		return false, nil
+	}
+	if p.flushingMemBuffer != nil {
+		if err := <-p.errCh; err != nil {
+			return false, err
+		}
+	}
+	p.onFlushing.Store(true)
+	p.flushingMemBuffer = p.memBuffer
+	p.len += p.flushingMemBuffer.Len()
+	p.size += p.flushingMemBuffer.Size()
+	p.memBuffer = newMemDB()
+	p.generation++
+	go func(generation uint64) {
+		util.EvalFailpoint("beforePipelinedFlush")
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.flushingMemBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.flushingMemBuffer.Size()))
+		err := p.flushFunc(generation, p.flushingMemBuffer)
+		p.onFlushing.Store(false)
+		// Send the error to errCh after onFlushing status is set to false.
+		// this guarantees the onFlushing.Store(true) in another goroutine's MayFlush happens after onFlushing.Store(false) in this function.
+		p.errCh <- err
+	}(p.generation)
+	return true, nil
+}
+
+func (p *PipelinedMemDB) needFlush() bool {
+	size := p.memBuffer.Size()
+	// size < MinFlushSize, do not flush.
+	// MinFlushSize <= size < MaxFlushSize && keys < MinFlushKeys, do not flush.
+	// MinFlushSize <= size < MaxFlushSize && keys >= MinFlushKeys, flush.
+	// size >= MaxFlushSize, flush.
```

Comment:
`MaxFlushSize` implies a maximum size limit for a flush. However, its true purpose is to act as a threshold to trigger a flush. A more obvious name could be something like `ForceFlushSizeThreshold`.